Sources/LeafKit/LeafData/LeafDataStorage.swift

```diff
@@ -3,46 +3,220 @@
 
 import Foundation
 
-internal enum LeafDataStorage {
-    /// A `Bool`.
-    ///
-    ///     true
-    ///
+internal indirect enum LeafDataStorage: Equatable, CustomStringConvertible {
+    // MARK: - Cases
+    
+    // Static values
     case bool(Bool)
-
-    /// A `String`.
-    ///
-    ///     "hello"
-    ///
     case string(String)
-
-    /// An `Int`.
-    ///
-    ///     42
-    ///
     case int(Int)
-
-    /// A `Double`.
-    ///
-    ///     3.14
-    ///
     case double(Double)
-
-    /// `Data` blob.
-    ///
-    ///     Data([0x72, 0x73])
-    ///
     case data(Data)
-
-    /// A nestable `[String: LeafData]` dictionary.
+    
+    // Collections (potentially holding lazy values)
     case dictionary([String: LeafData])
-
-    /// A nestable `[LeafData]` array.
     case array([LeafData])
 
-    /// A lazily-resolvable `LeafData`.
-    case lazy(() -> (LeafData))
+    // Wrapped `Optional<LeafDataStorage>`
+    case optional(_ wrapped: LeafDataStorage?, _ type: LeafData.NaturalType)
+    
+    // Lazy resolvable function
+    // Must specify return tuple giving (returnType, invariance)
+    case lazy(f: () -> (LeafData),
+              returns: LeafData.NaturalType,
+              invariant: Bool)
```

Comment:
* fix comment, signature updated:
f - the LeafData generator
returns: the directed natural type (eg, .string)
invariant: whether the lazy generator can safely be called at any time with the same results

EX: a lazy generator might return the computer's IP address. This *might* be invariant but might not be, so care has to be taken, as an *invariant* generator will be *RESOLVED IMMEDIATELY* and not at serialization.

Thus, such a function would require careful handling.